File: 백준/Gold/2073. 수도배관공사/수도배관공사.py
import sys
input = sys.stdin.readline

d, p = map(int, input().split())
pipe = []
for _ in range(p):
    l, c = map(int, input().split())
    pipe.append((l,c))

INF = sys.maxsize
dp = [INF] + [0]*d

for l, c in pipe:
    pre_dp = dp.copy()
    for cur_l in range(l, d+1):
        if pre_dp[cur_l - l]:
            dp[cur_l] = max(dp[cur_l], min(pre_dp[cur_l-l], c))

print(dp[-1])

File: 백준/Gold/2073. 수도배관공사/수도배관공사.js
const input = require('fs').readFileSync(process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt').toString().split('\n')

const [D, P] = input[0].split(' ').map(Number)
const pipes = []
for (let i = 1; i < P + 1; i++) {
  pipes.push(input[i].split(' ').map(Number))
}

const dp = [Infinity, ...Array(D).fill(0)]
for (const [L, C] of pipes) {
  for (let l = D; l >= L; l--) {
    dp[l] = Math.max(dp[l], Math.min(dp[l - L], C))
  }
}
console.log(dp[D])